test/lint/lint-files.py

```diff
@@ -0,0 +1,207 @@
+        """
+        Returns the octal file permission of the file
+        """
+        return int(oct(os.stat(self.file_path).st_mode)[-3:])
+
+
+def check_all_filenames() -> int:
+    """
+    Checks every file in the repository against an allowed regexp to make sure only lowercase or uppercase
+    alphanumerics (a-zA-Z0-9), underscores (_), hyphens (-), at (@) and dots (.) are used in repository filenames.
+    """
+    # We avoid using rstrip() to ensure we catch filenames which accidentally include trailing whitespace
+    filenames = check_output(CMD_ALL_FILES, shell=True).decode("utf8").split("\n")
+    filenames = [filename for filename in filenames if filename != ""]  # removes the trailing empty list element
+
+    filename_regex = re.compile(ALLOWED_FILENAME_REGEXP)
+    failed_tests = 0
+    for filename in filenames:
+        if not filename_regex.match(filename):
+            print(
+                f"""File "{filename}" does not not match the allowed filename regexp ('{ALLOWED_FILENAME_REGEXP}')."""
+            )
+            failed_tests += 1
+    return failed_tests
+
+
+def check_source_filenames() -> int:
+    """
+    Checks only source files (*.cpp, *.h, *.py, *.sh) against a stricter allowed regexp to make sure only lowercase
+    alphanumerics (a-z0-9), underscores (_), hyphens (-) and dots (.) are used in source code filenames.
+
+    Additionally there is an exception regexp for directories or files which are excepted from matching this regexp.
+    """
+    # We avoid using rstrip() to ensure we catch filenames which accidentally include trailing whitespace
+    filenames = check_output(CMD_SOURCE_FILES, shell=True).decode("utf8").split("\n")
+    filenames = [filename for filename in filenames if filename != ""]  # removes the trailing empty list element
+
+    filename_regex = re.compile(ALLOWED_SOURCE_FILENAME_REGEXP)
+    filename_exception_regex = re.compile(ALLOWED_SOURCE_FILENAME_EXCEPTION_REGEXP)
+    failed_tests = 0
+    for filename in filenames:
+        if not filename_regex.match(filename) and not filename_exception_regex.match(filename):
+            print(
+                f"""File "{filename}" does not not match the allowed source filename regexp ('{ALLOWED_SOURCE_FILENAME_REGEXP}'), or the exception regexp ({ALLOWED_SOURCE_FILENAME_EXCEPTION_REGEXP})."""
+            )
+            failed_tests += 1
+    return failed_tests
+
+
+def check_all_file_permissions() -> int:
+    """
+    Checks all files in the repository match an allowed executable or non-executable file permission octal.
+
+    Additionally checks that for executable files, the file contains a shebang line
+    """
+    filenames = check_output(CMD_ALL_FILES, shell=True).decode("utf8").strip().split("\n")
+    failed_tests = 0
+    for filename in filenames:
+        file_meta = FileMeta(filename)
+        if file_meta.permissions == ALLOWED_PERMISSION_EXECUTABLES:
+            shebang = open(filename, "rb").readline().rstrip(b"\n")
```

Comment:
nit: shouldn't `open` be used in a context manager, so that it is closed again?

Reply:
Good call, I will address in a follow-up